### PR TITLE
*: support generate MySQL/MariaDB binlog events for testing

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

ref #34 

### What is changed and how it works?

- add functions to support generate most important binlog events
- add a wrapper to support generate random binlog events/files

---

- 3 levels to generate binlog events (lowest to hightest)
	1. generate specified binlog events (`event.go`)
	2. generate binlog events for file header (`common.go`), DDL (`ddl.go`) and DML (`dml.go`)
	3. generate binlog events for file (`generator.go`)
- for the usage, see unit test cases

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@GregoryIan @amyangfei PTAL

**Comment by GregoryIan** [comment]
Good Job！

File: pkg/binlog/event/common.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	)
+
+	switch flavor {
+	case gmysql.MySQLFlavor:
+		uuidSet := singleGTID.(*gmysql.UUIDSet)
+		interval := uuidSet.Intervals[0]
+		gtidEv, err = GenGTIDEvent(header, latestPos, defaultGTIDFlags, uuidSet.SID.String(), interval.Start, defaultLastCommitted, defaultSequenceNumber)
+	case gmysql.MariaDBFlavor:
+		mariaGTID := singleGTID.(*gmysql.MariadbGTID)
+		if mariaGTID.ServerID != header.ServerID {
```
Comment (GregoryIan) [inline pkg/binlog/event/common.go]: lol

File: pkg/binlog/event/common.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+	case gmysql.MySQLFlavor:
+		mysqlGTIDs, ok := origin.(*gmysql.MysqlGTIDSet)
+		if !ok {
+			return nil, errors.NotValidf("GTID set string %s for MySQL", gSet)
+		}
+		if len(mysqlGTIDs.Sets) != 1 {
+			return nil, errors.Errorf("only one GTID in set is supported, but got %d (%s)", len(mysqlGTIDs.Sets), gSet)
+		}
+		var uuidSet *gmysql.UUIDSet
+		for _, uuidSet = range mysqlGTIDs.Sets {
```
Comment (GregoryIan) [inline pkg/binlog/event/common.go]: remove it？

File: pkg/binlog/event/common.go (hunk 1 above)
Comment (csuzhangxc) [reply pkg/binlog/event/common.go]: for MariaDB, GTID including the `server-id`. :relieved:

File: pkg/binlog/event/event.go [hunk 3]
```diff
@@ -0,0 +1,759 @@
...
+
+// GenEventHeader generates a EventHeader's raw data according to a passed-in EventHeader struct.
+// ref: https://dev.mysql.com/doc/internals/en/binlog-event-header.html
+func GenEventHeader(header *replication.EventHeader) ([]byte, error) {
+	buf := new(bytes.Buffer)
+
+	// timestamp, 4 bytes
+	err := binary.Write(buf, binary.LittleEndian, header.Timestamp)
+	if err != nil {
+		return nil, errors.Annotatef(err, "write timestamp %d", header.Timestamp)
```
Comment (GregoryIan) [inline pkg/binlog/event/event.go]: do you want add "[generate event header]" information in the error?

File: pkg/binlog/event/util.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	if err != nil {
+		return nil, errors.Annotatef(err, "decode % X", buf.Bytes())
+	}
+
+	return &replication.BinlogEvent{RawData: buf.Bytes(), Header: &header, Event: event}, nil
+}
+
+// combineHeaderPayload combines header, postHeader and payload together.
+func combineHeaderPayload(buf *bytes.Buffer, header, postHeader, payload []byte) error {
+	err := binary.Write(buf, binary.LittleEndian, header)
```
Comment (GregoryIan) [inline pkg/binlog/event/util.go]: do you want check `its length == eventHeaderLen`?

File: pkg/binlog/event/event.go (hunk 3 above)
Comment (csuzhangxc) [reply pkg/binlog/event/event.go]: I tend to:

if `A` call `B`, `B` call `C`, and `C` return an error.
then
1. annotate `err` in `B` with something like `calling C fail`
2. annotate `err` in `A` with something like `calling B fail`

use this way, if `B` also call `D` and `D` return an error, then only need annotate `err` with something like `calling D` fail.

so, in this L66, I only annotate the error with information about L64, and annotate something like `generate event header` in the outer function.

File: pkg/binlog/event/common.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+//   -. MariadbBinlogCheckPointEvent, not added yet
+func GenCommonFileHeader(flavor string, serverID uint32, gSet gtid.Set) ([]*replication.BinlogEvent, []byte, error) {
+	var (
+		header = &replication.EventHeader{
+			Timestamp: uint32(time.Now().Unix()),
+			ServerID:  serverID,
+			Flags:     defaultHeaderFlags,
+		}
+		latestPos       = uint32(len(replication.BinLogFileHeader))
+		prevGTIDsEv     *replication.BinlogEvent // for MySQL, this will be nil
```
Comment (GregoryIan) [inline pkg/binlog/event/common.go]: not implemented?

File: pkg/binlog/event/common.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+			Flags:     defaultHeaderFlags,
+		}
+		gtidEv *replication.BinlogEvent
+	)
+
+	switch flavor {
+	case gmysql.MySQLFlavor:
+		uuidSet := singleGTID.(*gmysql.UUIDSet)
+		interval := uuidSet.Intervals[0]
+		gtidEv, err = GenGTIDEvent(header, latestPos, defaultGTIDFlags, uuidSet.SID.String(), interval.Start, defaultLastCommitted, defaultSequenceNumber)
```
Comment (GregoryIan) [inline pkg/binlog/event/common.go]: use `interval.Start`?

File: pkg/binlog/event/common.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+// for MySQL:
+//   1. BinLogFileHeader, [ fe `bin` ]
+//   2. FormatDescriptionEvent
+//   3. PreviousGTIDsEvent
+// for MariaDB:
+//   1. BinLogFileHeader, [ fe `bin` ]
+//   2. FormatDescriptionEvent
+//   3. MariadbGTIDListEvent
+//   -. MariadbBinlogCheckPointEvent, not added yet
+func GenCommonFileHeader(flavor string, serverID uint32, gSet gtid.Set) ([]*replication.BinlogEvent, []byte, error) {
```
Comment (GregoryIan) [inline pkg/binlog/event/common.go]: for `Genxxx` functions,  I suggest to add `latestPos` for all them

File: pkg/binlog/event/common.go (hunk 6 above)
Comment (GregoryIan) [reply pkg/binlog/event/common.go]: ok, I know  but still confused first time

File: pkg/binlog/event/common.go [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+	singleGTID, err := verifySingleGTID(flavor, gSet)
+	if err != nil {
+		return nil, errors.Annotatef(err, "verify single GTID in set")
+	}
+	clone := gSet.Clone()
+
+	switch flavor {
+	case gmysql.MySQLFlavor:
+		uuidSet := singleGTID.(*gmysql.UUIDSet)
+		uuidSet.Intervals[0].Start++
```
Comment (GregoryIan) [inline pkg/binlog/event/common.go]: clone `uuidSet`

File: pkg/binlog/event/common.go [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+		uuidSet.Intervals[0].Stop++
+		gtidSet := new(gmysql.MysqlGTIDSet)
+		gtidSet.Sets = map[string]*gmysql.UUIDSet{uuidSet.SID.String(): uuidSet}
+		err = clone.Set(gtidSet)
+	case gmysql.MariaDBFlavor:
+		mariaGTID := singleGTID.(*gmysql.MariadbGTID)
+		mariaGTID.SequenceNumber++
```
Comment (GregoryIan) [inline pkg/binlog/event/common.go]: ditto

File: pkg/binlog/event/common.go (hunk 7 above)
Comment (csuzhangxc) [reply pkg/binlog/event/common.go]: the header has no meaningful pos/latestPos
1. 4 bytes magic num
2. (123-4) bytes FormatDescriptionEvent (currently with CRC32 checksum enabled)
3. variable length PreviousGTIDsEvent/MariadbGTIDListEvent

File: pkg/binlog/event/common.go (hunk 8 above)
Comment (csuzhangxc) [reply pkg/binlog/event/common.go]: already cloned in `gSet.Origin()` call in `verifySingleGTID`.

File: pkg/binlog/event/common.go (hunk 9 above)
Comment (csuzhangxc) [reply pkg/binlog/event/common.go]: same above

File: pkg/binlog/event/common.go (hunk 7 above)
Comment (GregoryIan) [reply pkg/binlog/event/common.go]: while I saw this code, I had to confirm the position mechanism, although it is correct, but not uniform

File: pkg/binlog/event/util.go (hunk 4 above)
Comment (csuzhangxc) [reply pkg/binlog/event/util.go]: added

File: pkg/binlog/event/common.go (hunk 7 above)
Comment (csuzhangxc) [reply pkg/binlog/event/common.go]: yeh, but if add a `latestPos`, then do not know which value should be passed.:worried:

**Comment by GregoryIan** [comment]
Rest LGTM

File: pkg/binlog/event/event.go (hunk 3 above)
Comment (GregoryIan) [reply pkg/binlog/event/event.go]: ok. Just be unified

File: pkg/binlog/event/common.go (hunk 7 above)
Comment (GregoryIan) [reply pkg/binlog/event/common.go]: we would use like this way
```
ev = GenXX(latestPos)
updateLatePos(ev)
```
so it's not a problem, expect diving into it

but it's not a big problem

File: pkg/binlog/event/generator.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	g.updateLatestPosGTID(result.LatestPos, result.LatestGTID)
+	return result.Events, result.Data, nil
+}
+
+// GenDDLEvents generates binlog events for DDL statements.
+// events: [GTIDEvent, QueryEvent]
+func (g *Generator) GenDDLEvents(schema string, query string) ([]*replication.BinlogEvent, []byte, error) {
```
Comment (GregoryIan) [inline pkg/binlog/event/generator.go]: I perfer it

Comment (csuzhangxc) [reply pkg/binlog/event/generator.go]: yeh. In fact, only this is enough for DDL.

File: pkg/binlog/event/common.go (hunk 7 above)
Comment (csuzhangxc) [reply pkg/binlog/event/common.go]: check latest GTID whether is one of the last GTID in previous GTIDs?

Comment (GregoryIan) [reply pkg/binlog/event/common.go]: yep

**Comment by GregoryIan** [comment]
LGTM

File: pkg/binlog/event/event.go [hunk 11]
```diff
@@ -0,0 +1,759 @@
...
+	eventHeaderLen         = uint8(replication.EventHeaderSize) // always 19
+	crc32Len        uint32 = 4                                  // CRC32-length
+	tableMapFlags   uint16 = 1                                  // flags in TableMapEvent's post-header, not used yet
+)
+
+var (
+	// A array indexed by `Binlog-Event-Type - 1` to extract the length of the event specific header.
+	// It is copied from a binlog file generated by MySQL 5.7.22-log.
+	// The doc at https://dev.mysql.com/doc/internals/en/format-description-event.html does not include all of them.
+	eventTypeHeaderLen = []byte{0x38, 0x0d, 0x00, 0x08, 0x00, 0x12, 0x00, 0x04, 0x04, 0x04, 0x04, 0x12, 0x00, 0x00, 0x5f, 0x00, 0x04, 0x1a, 0x08, 0x00, 0x00, 0x00, 0x08, 0x08, 0x08, 0x02, 0x00, 0x00, 0x00, 0x0a, 0x0a, 0x0a, 0x2a, 0x2a, 0x00, 0x12, 0x34, 0x00}
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/binlog/event/event.go]: split to more lines (every line with 16 bytes)

File: pkg/binlog/event/util.go [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+	return int(columnCount+7) / 8
+}
+
+// nullBytes returns a n-length null bytes slice
+func nullBytes(n int) []byte {
+	buf := new(bytes.Buffer)
+	for i := 0; i < n; i++ {
+		buf.WriteByte(0x00)
+	}
+	return buf.Bytes()
```
Comment (amyangfei) [inline pkg/binlog/event/util.go]: we can get a n-length null bytes slice by `make([]byte, n)` directly?

File: pkg/binlog/event/common.go [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+	case gmysql.MariaDBFlavor:
+		prevGTIDsEv, err = GenMariaDBGTIDListEvent(header, latestPos, gSet)
+		if err == nil {
+			prevGTIDsEvData = prevGTIDsEv.RawData
+		}
+	default:
+		return nil, nil, errors.NotSupportedf("flavor %s", flavor)
+	}
+	if err != nil {
+		return nil, nil, errors.Annotatef(err, "generate PreviousGTIDsEvent/MariadbGTIDListEvent")
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/binlog/event/common.go]: ❤️ addressed.

File: pkg/binlog/event/util.go (hunk 12 above)
Comment (csuzhangxc) [reply pkg/binlog/event/util.go]: yeh, we can

**Pull request merged**